quinn/src/delay_queue.rs

```diff
@@ -0,0 +1,442 @@
+                "head of list has no predecessor"
+            );
+        }
+        self.list_unlink(timer);
+    }
+
+    /// Remove `timer` from its list
+    fn list_unlink(&mut self, timer: Timer) {
+        if let Some(prev) = self.timers[timer.0].prev.take() {
+            // Remove reference from predecessor
+            self.timers[prev.0].next = self.timers[timer.0].next;
+        }
+        if let Some(next) = self.timers[timer.0].next.take() {
+            // Remove reference from successor
+            self.timers[next.0].prev = self.timers[timer.0].prev;
+        }
+    }
+}
+
+fn range_in_level(level: usize, raw: RangeInclusive<u64>) -> RangeInclusive<usize> {
+    let shift = level * LOG_2_SLOTS;
+    const MASK: u64 = SLOTS as u64 - 1;
+    let start = ((*raw.start() >> shift) & MASK) as usize;
+    let level_end = (*raw.start() >> shift) | MASK;
+    let end = ((*raw.end() >> shift).min(level_end) & MASK) as usize;
+    start..=end
+}
+
+/// Compute the first tick that lies within a slot
+fn slot_start(base: u64, level: usize, slot: usize) -> u64 {
+    let shift = (level * LOG_2_SLOTS) as u64;
+    (base & ((!0 << shift) << LOG_2_SLOTS as u64)) | ((slot as u64) << shift)
+}
+
+/// Compute the level and slot for a certain expiry
+fn timer_index(base: u64, expiry: u64) -> (usize, usize) {
+    let differing_bits = base ^ expiry;
+    let level = (63 - (differing_bits | 1).leading_zeros()) as usize / LOG_2_SLOTS;
+    debug_assert!(level < LEVELS, "every possible expiry is in range");
+    let slot_base = (base >> (level * LOG_2_SLOTS)) & (!0 << LOG_2_SLOTS);
+    let slot = (expiry >> (level * LOG_2_SLOTS)) - slot_base;
+    debug_assert!(slot < SLOTS as u64);
+    (level, slot as usize)
+}
+
+impl<T> Default for DelayQueue<T> {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+#[derive(Debug)]
+struct TimerState<T> {
+    /// Lowest argument to `poll` for which this timer may be returned
+    expiry: u64,
+    /// Value returned to the caller on expiry
+    value: T,
+    /// Predecessor within a slot's list
+    prev: Option<Timer>,
+    /// Successor within a slot's list
+    next: Option<Timer>,
+}
+
+/// A set of contiguous timer lists, ordered by expiry
+#[derive(Copy, Clone)]
+struct Level {
+    slots: [Option<Timer>; SLOTS],
+}
+
+impl Level {
+    fn new() -> Self {
+        Self {
+            slots: [None; SLOTS],
+        }
+    }
+}
+
+impl fmt::Debug for Level {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let mut m = f.debug_map();
+        for (i, Timer(t)) in self
+            .slots
+            .iter()
+            .enumerate()
+            .filter_map(|(i, x)| x.map(|t| (i, t)))
+        {
+            m.entry(&i, &t);
+        }
+        m.finish()
+    }
+}
+
+const LEVELS: usize = 11;
+const LOG_2_SLOTS: usize = 6;
+const SLOTS: usize = 1 << LOG_2_SLOTS;
+
+// Future work: add a niche here
+#[derive(Debug, Copy, Clone, Eq, PartialEq)]
+pub struct Timer(usize);
```

Comment:
maybe add a comment that the `usize` is an index into the slab. When I looked at this before checking how `Timer` is actually used I assumed it would be a timestamp.

Reply:
Done.